sklearn/metrics/_pairwise_distances_reduction/_gemm_term_computer.pxd.tp

```diff
@@ -76,6 +76,34 @@ cdef class GEMMTermComputer{{name_suffix}}:
         ITYPE_t thread_num
     ) nogil
 
+
+cdef class DenseDenseMiddleTermComputer{{name_suffix}}(MiddleTermComputer{{name_suffix}}):
+    cdef DTYPE_t * _compute_distances_on_chunks(
+        self,
+        ITYPE_t X_start,
+        ITYPE_t X_end,
+        ITYPE_t Y_start,
+        ITYPE_t Y_end,
+        ITYPE_t thread_num,
+    ) nogil
+
+
+cdef void _middle_term_sparse_sparse_{{name_suffix}}(
+    const {{INPUT_DTYPE_t}}[:] X_data,
+    const ITYPE_t[:] X_indices,
+    const ITYPE_t[:] X_indptr,
+    ITYPE_t X_start,
+    ITYPE_t X_end,
+    const {{INPUT_DTYPE_t}}[:] Y_data,
+    const ITYPE_t[:] Y_indices,
+    const ITYPE_t[:] Y_indptr,
+    ITYPE_t Y_start,
+    ITYPE_t Y_end,
+    DTYPE_t[:, ::1] D
+) nogil
```

Comment:
Conventionally, utility functions' are declared before classes: this prevents readers from mistaking a function for a method.